src/runtime/parcelset/parcel.cpp

```diff
@@ -5,75 +5,247 @@
+        if (HPX_UNLIKELY(!components::types_are_compatible(
+            data_.addr_.type_, comptype)))
+        {
+            std::ostringstream strm;
+            strm << " types are not compatible: destination_type("
+                  << data_.addr_.type_ << ") action_type(" << comptype
+                  << ") parcel ("  << *this << ")";
+            HPX_THROW_EXCEPTION(bad_component_type,
+                "applier::schedule_action",
+                strm.str());
+        }
+
+        return lva;
     }
 
-    void parcel::serialize(serialization::input_archive & ar, unsigned)
+    void parcel::load_schedule(serialization::input_archive & ar,
+        std::size_t num_thread)
     {
-        ar & data_;
-        if(data_.has_source_id_)
+        using hpx::serialization::detail::polymorphic_id_factory;
+
+        ar >> data_;
+        ar >> cont_;
+
+        std::uint32_t id;
+        ar >> id;
+#if !defined(HPX_DEBUG)
+        action_.reset(polymorphic_id_factory::create<actions::base_action>(id));
+#else
+        std::string name;
+        ar >> name;
+        action_.reset(
+            polymorphic_id_factory::create<actions::base_action>(id, &name));
+#endif
+        // make sure this parcel destination matches the proper locality
+        HPX_ASSERT(destination_locality() == data_.addr_.locality_);
+
+        naming::address_type lva = determine_lva();
+
+        // make sure the target has not been migrated away
+        auto r = action_->was_object_migrated(data_.dest_, lva);
+        if (r.first)
         {
-            ar & source_id_;
+            naming::resolver_client& client = hpx::naming::get_agas_client();
+            // If the object was migrated, just load the action and route.
+            ar >> *action_;
+            client.route(
+                std::move(*this),
+                &detail::parcel_sent_handler,
+                threads::thread_priority_normal);
+            return;
+        }
+
+        // dispatch action, register work item either with or without
+        // continuation support
+        if (cont_)
+        {
+            // No continuation is to be executed, register the plain
+            // action and the local-virtual address.
+            action_->load_schedule(ar, std::move(cont_), std::move(data_.dest_),
+                lva, num_thread);
         }
         else
         {
-            source_id_ = naming::invalid_id;
+            // This parcel carries a continuation, register a wrapper
+            // which first executes the original thread function as
+            // required by the action and triggers the continuations
+            // afterwards.
+            action_->load_schedule(ar, std::move(data_.dest_), lva, num_thread);
         }
-        ar & dests_;
-        ar & addrs_;
-        ar & cont_;
-        ar & action_;
     }
 
-    void parcel::serialize(serialization::output_archive & ar, unsigned)
+    void parcel::schedule_action()
     {
```

Comment:
Am 02.09.2016 6:02 nachm. schrieb "Hartmut Kaiser" <notifications@github.com
>:
>
> In src/runtime/parcelset/parcel.cpp:
>
> > } > > - void parcel::serialize(serialization::output_archive & ar,
unsigned) > + void parcel::schedule_action() > {
>
> Are direct_actions handled now in special ways on the receiving end?

Yes. They are just a function call now.

>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.